Memory snapshot harness havoc

xbauch opened the pull request:
This PR adds the option to selectively initialise global variables from a snapshot to a non-deterministic value rather than the value they had in the snapshot.

The scope of this PR are global variables of primitive types but since we use the `recursive_initialization` for the havoc-ing, the scope is inherited from the goto-harness general. At this moment, `struct_tag`, `pointer`, and `array` have dedicated treatment. On the other hand it is not within the scope of this PR to modify the configuration of the recursive initialization, i.e. the default values are used.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

xbauch commented:
Only the actual havoc-ing commit (for early feedback), documentation and tests are coming.

sonodtt commented on doc/cprover-manual/goto-harness.md:
```diff
@@ -309,4 +310,97 @@ main.c function function
+The `function-call` harness is used in situations when we want the analysed
+function to work in arbitrary environment. If we want a more restricted
+environment or if we have the program in which our function will only be called,
+we call the `memory-snapshot` harness instead.
```
we call -> we can call

xbauch replied:
Done.

sonodtt commented on doc/cprover-manual/goto-harness.md:
```diff
@@ -309,4 +310,97 @@ main.c function function
+
+### The memory snapshot harness
+
+The `function-call` harness is used in situations when we want the analysed
```
situations when -> situations in which

xbauch replied:
Done.

sonodtt commented on doc/cprover-manual/goto-harness.md:
```diff
@@ -309,4 +310,97 @@ main.c function function
+
+The `function-call` harness is used in situations when we want the analysed
+function to work in arbitrary environment. If we want a more restricted
+environment or if we have the program in which our function will only be called,
```
remove 'only'   function will only be called

xbauch replied:
I wanted "only" there somewhere. What about:
..or if we have the only program in which our function will be called..
?

sonodtt commented on doc/cprover-manual/goto-harness.md:
```diff
@@ -309,4 +310,97 @@ main.c function function
+environment or if we have the program in which our function will only be called,
+we call the `memory-snapshot` harness instead.
+
+Furthermore, the program state of interest may be taken a particular location
```
remove 'taken' ?  might be clear enough. Alternatively - 'taken from'

xbauch replied:
What about "taken at a particular location"?

sonodtt commented on doc/cprover-manual/goto-harness.md:
````diff
@@ -309,4 +310,97 @@ main.c function function
+}
+```
+
+But are not particularly interested in the analysing the complex function, since
````
'interested in the'   -> 'interested in analysing'

pps sorry about all the minor fixes. 100% aware english is not your 1st language - and you're doing great.

xbauch replied:
Done. And no reason to apologise, the language is important especially in the manual.

hannes-steffenhagen-diffblue commented on src/goto-harness/memory_snapshot_harness_generator.h:
```diff
@@ -21,7 +21,8 @@ Author: Daniel Poetzl
   "(memory-snapshot):"                                                         \
-  "(initial-location):" // MEMORY_SNAPSHOT_HARNESS_GENERATOR_OPTIONS
+  "(initial-location):"                                                        \
+  "(havoc-variables):" // MEMORY_SNAPSHOT_HARNESS_GENERATOR_OPTIONS
```
This section should have `// clang-format off` and `// clang-format on` options to prevent clang-format from messing up the formatting here

xbauch replied:
Done.

hannes-steffenhagen-diffblue commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -157,18 +163,29 @@ void memory_snapshot_harness_generatort::add_init_section(
   code_blockt &code) const
 {
+  recursive_initialization_configt recursive_initialization_config;
+  auto recursive_initialization = util_make_unique<recursive_initializationt>(
```
If you're just using this in-place here there's no need for dynamic allocation... we only had that in the other harness generator because it was a stateful mess

xbauch replied:
Done.

hannes-steffenhagen-diffblue commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -54,6 +56,10 @@ void memory_snapshot_harness_generatort::handle_option(
       location_number = optionalt<unsigned>(safe_string2unsigned(start.back()));
     }
   }
+  else if (option == "havoc-variables")
```
s/variables/globals?

hannes-steffenhagen-diffblue commented on doc/cprover-manual/goto-harness.md:
```diff
@@ -309,4 +310,97 @@ main.c function function
+
+### The memory snapshot harness
+
+The `function-call` harness is used in situations when we want the analysed
```
I find this sentence to be a bit awkward. The point of the snapshot harness is to analyse a function starting from a "real" program state (or a subset thereof)

xbauch replied:
Changed.

hannes-steffenhagen-diffblue commented on doc/cprover-manual/goto-harness.md:
```diff
@@ -309,4 +310,97 @@ main.c function function
+Furthermore, the program state of interest may be taken a particular location
+within a function. In that case we do not want the harness to instrument the
+whole function but rather to allow starting the execution from a specific
+initial location (specified via `--initial-location func[:<n>]`).
```
Might be worth noting here that the `--initial-location` here should match the program location at the original breakpoint location

xbauch replied:
Added.

NlightNFotis commented on doc/cprover-manual/goto-harness.md:
```diff
@@ -309,4 +310,97 @@ main.c function function
+But are not particularly interested in the analysing the complex function, since
+we trust that its implementation is correct. Hence we run the above program
+stopping after the assignments to `x` and `x` and storing the program state,
+e.g. using the `memory-analyzer`, in a JSON file `snapshot.json`. Then run the
```
Do we have any documentation for the `memory-analyzer`? I recall seeing some around, but I can't seem to be able to find it now? If we don't, would it be a good idea to write some separate documentation for that as well?

xbauch replied:
https://github.com/xbauch/cbmc/pull/3 I intent to make it part of #4261 once review-able.

NlightNFotis reviewed (commented):
This looks good to me overall.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: aeff777284d4f7e3cda57d9968869d0ecc9ad598).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/104250683

tautschnig commented on regression/goto-harness/havoc-global-int-01/test.desc:
```diff
@@ -0,0 +1,8 @@
+^SIGNAL=0$
+\[main.assertion.1\] line [0-9]+ assertion x == 1: FAILURE
+--
+^warning: ignoring
```
I think it would be helpful to document how the JSON snapshots were generated. There will be a day where we need to change them in some way.

xbauch replied:
That documentation was intended to be a part of #4261 (as we used the memory-analyzer to generate the JSON). Do you think it should be documented here as well?

tautschnig commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -157,18 +162,29 @@ void memory_snapshot_harness_generatort::add_init_section(
+    if(variables_to_havoc.count(symbol.base_name) == 0)
+    {
+      code_assignt code_assign(symbol.symbol_expr(), symbol.value);
+      code.add(code_assign);
```
Nit picking: I don't think the `code_assign` temporary adds value, you might just do `code.add(code_assignt{...});`. If you do prefer to keep it I'd suggest to use `code.add(std::move(code_assign));`

xbauch replied:
Done.

tautschnig commented on src/goto-harness/memory_snapshot_harness_generator.h:
```diff
@@ -48,30 +49,92 @@ class memory_snapshot_harness_generatort : public goto_harness_generatort
   {
   }
 
+  /// 1. load memory table from the snapshot
```
I'd start this documentation with a one-sentence summary. Otherwise I'd suspect this to render awkwardly.

xbauch replied:
Done.

tautschnig commented on src/goto-harness/memory_snapshot_harness_generator.h:
```diff
@@ -48,30 +49,92 @@ class memory_snapshot_harness_generatort : public goto_harness_generatort
   }
 
+  /// 1. load memory table from the snapshot
+  /// 2. add initial section to the initial location
```
Also I think it would be nice to make these proper sentences, starting with an uppercase letter and terminated with a full stop.

xbauch replied:
Done.

tautschnig commented on src/goto-harness/memory_snapshot_harness_generator.h:
```diff
@@ -48,30 +49,92 @@ class memory_snapshot_harness_generatort : public goto_harness_generatort
 
+  /// Parse the snapshot JSON file and initialise the symbol table
+  /// \param file: the snapshot JSON file
+  /// \param snapshot: the resulting symbol table build from the snapshot
```
s/build/built/

xbauch replied:
Done.

tautschnig commented on src/goto-harness/memory_snapshot_harness_generator.h:
```diff
@@ -48,30 +49,92 @@ class memory_snapshot_harness_generatort : public goto_harness_generatort
   void
   get_memory_snapshot(const std::string &file, symbol_tablet &snapshot) const;
 
+  /// Turn this:
```
I'd start with a one-sentence summary. Otherwise this likely renders awkwardly.

xbauch replied:
Done.

tautschnig commented on src/goto-harness/memory_snapshot_harness_generator.h:
```diff
@@ -48,30 +49,92 @@ class memory_snapshot_harness_generatort : public goto_harness_generatort
 
+  /// For each global symbol in the \p snapshot symbol table either:
+  /// 1) add \ref code_assignt assigning a value from the \p snapshot to the
+  ///    symbol
```
add "or"

xbauch replied:
Done.

tautschnig commented on doc/cprover-manual/goto-harness.md:
```diff
@@ -309,4 +310,102 @@ main.c function function
+
+The `function-call` harness is used in situations in which we want the analysed
+function to work in arbitrary environment. If we want a more restricted
+environment or if we have the program in which our function will be called,
```
or if we have an executable program in which

xbauch replied:
Done.

tautschnig commented on doc/cprover-manual/goto-harness.md:
```diff
@@ -309,4 +310,102 @@ main.c function function
+initial location (specified via `--initial-location func[:<n>]`). Note that the
+initial location does not have to be the first instruction of a function: we can
+also specify the _location number_ `n` to set the initial location inside our
+function. The _location numbers_ do not have to co-relate with the lines of the
```
s/co-relate/correlate/ (or maybe "coincide")

xbauch replied:
Done.

hannes-steffenhagen-diffblue commented on regression/goto-harness/havoc-global-int-01/main.c:
```diff
@@ -0,0 +1,8 @@
+
+int main()
+{
+  assert(x == 1);
```
what exactly is this test testing? Presumably this assertion would still fail if the snapshot harness _didn't_ work?

xbauch replied:
Modified.

hannes-steffenhagen-diffblue commented on regression/goto-harness/havoc-global-int-02/main.c:
```diff
@@ -0,0 +1,52 @@
+  fill_array(large_array, array_size);
+  y = array_sum(large_array, array_size);
+  checkpoint();
+  assert(y + 2 > x);
```
can you have an `assert(false)` after `checkpoint` just to check that the generated code actually makes this part still reachable?

hannes-steffenhagen-diffblue reviewed (approved):
I think this looks reasonable, but I'd take another look at the regression tests

Commit:
Havoc global variables in memory snapshot harness

Basically just grabs a list of var-names from command line and calls recursive
initialize on them instead of assigning their value.

Commit:
Add documentation to memory snapshot harness

Code level documentation for the class methods.

Commit:
Add cprover manual entry

Covering memory-snapshot harness, but not the memory analyzer.

Commit:
Add regression tests

For global int, global struct, multiple snapshot-ed symbols, multiple havoc-ed
symbols.
